File: asg2-c-lex-sort/List.c
//  Tz-Shiuan Lin
//  1411593
//  List.c
//  CS101-pa2
//


#include "List.h"

typedef struct NodeObj{
    struct NodeObj *beforeNode;
    struct NodeObj *afterNode;
    int data;
} NodeObj;

typedef NodeObj* Node;

typedef struct ListObj{
    Node front;
    Node back;
    Node cursor;
    int length;
    int index;
} ListObj;


/* Constructors-Destructors */
List newList(void){
    List new_list = malloc(sizeof(ListObj));
    new_list->back = NULL;
    new_list->front = NULL;
    new_list->cursor = NULL;
    new_list->length = 0;
    new_list->index = -1;
    
    return new_list;
}
void freeList(List *pL){
    
    if(pL != NULL && *pL != NULL) {
        while (length(*pL) > 0) {
            moveFront(*pL);
            delete(*pL);
        }
        free(*pL);
    }
}

/* Access functions */
int length(List L){
    return L->length;
}

int index(List L){
    return L->index;
}

int front(List L){
    if (L->length <= 0) {
        printf("Empty_list_error : List is empty");
        exit(0);
    } else {
        return L->front->data;
    }
}

int back(List L){
    if (L->length <= 0) {
        printf("Empty_list_error : List is empty");
        exit(0);
    }
    return L->back->data;
}

int get(List L){
    if (L->length <= 0 && L->index < 0) {
        printf("Cursor_error : Cursor is not defined");
        exit(0);
    }
    return L->cursor->data;
}

int equals(List A, List B){
    if (A->length != B->length) {
        return 0;
    }
    moveFront(A);
    moveFront(B);
    while (index(A) != -1){
        if (get(A) == get(B)) {
            moveNext(A);
            moveNext(B);
        } else {
            return 0;
        }
    }
    return 1;
}

/* Manipulation procedures */
void clear(List L){
    while (length(L) > 0) {
        moveFront(L);
        delete(L);
    }
    L->cursor = NULL;
    L->back = NULL;
    L->front = NULL;
    L->length = 0;
    L->index = -1;
}

void moveFront(List L){
    if (L->length > 0) {
        L->cursor = L->front;
        L->index = 0;
    }
}

void moveBack(List L){
    L->cursor = L->back;
    L->index = L->length - 1;
}

void movePrev(List L){
    if (L->length < 0 || L->index == 0) {
        L->cursor = NULL;
        L->index = -1;
    } else {
        L->cursor = L->cursor->beforeNode;
        L->index = L->index -1;
    }
}
void moveNext(List L){
    if (L->length < 0 || L->index == L->length -1) {
        L->cursor = NULL;
        L->index = -1;
    } else {
        L->cursor = L->cursor->afterNode;
        L->index = L->index + 1;
    }
    
}
void prepend(List L, int data){
    Node newFront = malloc(sizeof(NodeObj));
    newFront->data = data;
    newFront->beforeNode = NULL;
    newFront->afterNode = L->front;
    if (L->length > 0) {
        L->front->beforeNode = newFront;
    } else {
        L->back = newFront;
    }
    L->front = newFront;
    if (L->index == 0) {
        L->cursor = newFront;
    } else if(L->index > 0){
        L->index = L->index + 1;
    }
    L->length = L->length + 1;
}
void append(List L, int data){
    Node newBack = malloc(sizeof(NodeObj));
    newBack->data = data;
    newBack->afterNode = NULL;
    newBack->beforeNode = L->back;
    if (L->length > 0) {
        L->back->afterNode = newBack;
    } else {
        L->front = newBack;
    }
    L->back = newBack;
    L->length = L->length + 1;
}
void insertBefore(List L, int data){
    if (L->index == -1) {
        printf("Cursor_error : Cursor is not defined");
        exit(0);
    }
    Node newBefore = malloc(sizeof(NodeObj));
    newBefore->afterNode = L->cursor;
    newBefore->beforeNode = L->cursor->beforeNode;
    newBefore->data = data;
    if (L->index != 0) {
        L->cursor->beforeNode->afterNode = newBefore;
    } else {
        L->front = newBefore;
    }
    L->cursor->beforeNode = newBefore;
    L->index = L->index + 1;
    L->length = L->length + 1;
}

void insertAfter(List L, int data){
    if (L->index == -1) {
        printf("Cursor_error : Cursor is not defined");
        exit(0);
    }
    Node newAfter = malloc(sizeof(NodeObj));
    newAfter->beforeNode = L->cursor;
    newAfter->afterNode = L->cursor->afterNode;
    newAfter->data = data;
    if (L->index == (L->length - 1)) {
        L->front = newAfter;
    }
    L->cursor->afterNode = newAfter;
    L->length = L->length + 1;
}
void deleteFront(List L){
    L->front->afterNode->beforeNode = NULL;
    L->front = L->front->afterNode;
    if (L->index == 0) {
        L->index = -1;
        L->cursor = NULL;
    }
    L->length = L->length - 1;
}
void deleteBack(List L){
    if (L->index == (L->length - 1)) {
        L->index = -1;
        L->cursor = NULL;
    }
    L->back->beforeNode->afterNode = NULL;
    L->back = L->back->beforeNode;
    L->length = L->length - 1;
}
void delete(List L){
    if (L->index == -1) {
        printf("Cursor_error : Cursor is not defined");
        exit(0);
    }
    Node before = NULL;
    Node after = NULL;
    if (L->cursor->beforeNode != NULL) {
        before = L->cursor->beforeNode;
    }
    if (L->cursor->afterNode) {
        after = L->cursor->afterNode;
    }
    if (before != NULL) {
        before->afterNode = after;
    }
    if (after != NULL) {
        after->beforeNode = before;
    }
    free(L->cursor);
    if (index(L) == 0) {
        L->front = after;
    }
    
    L->cursor = NULL;
    L->index = -1;
    L->length = L->length - 1;
}

/* Other operations */
void printList(FILE* out, List L){
    if(out == NULL){
        printf("Output file not found");
    }
    moveFront(L);
    while (L->index != -1) {
        fprintf(out, "%d ", L->cursor->data);
        moveNext(L);
    }
}

List copyList(List L){
    List newList = malloc(sizeof(ListObj));
    newList->back = NULL;
    newList->front = NULL;
    newList->cursor = NULL;
    newList->length = 0;
    newList->index = -1;

    Node oldNode = L->front;
    while (oldNode != NULL){
        append(newList, oldNode->data);
        oldNode = oldNode->afterNode;
    }
    return newList;
}

List concatList(List A, List B){
    List newList = (List)malloc(sizeof(List));
    Node newNode = malloc(sizeof(NodeObj));
    moveFront(A);
    newNode->data = A->cursor->data;
    newNode->afterNode = newNode->beforeNode = NULL;
    moveNext(A);
    while (A->index != -1) {
        append(newList, A->cursor->data);
        moveNext(A);
    }
    moveFront(B);
    while (B->index != -1) {
        append(newList, B->cursor->data);
        moveNext(B);
    }
    newList->cursor = NULL;
    newList->index = -1;
    return  newList;
}


/* NODE FUNCTION */
Node newNode(int data){
    Node newNode = malloc(sizeof(NodeObj));
    return newNode;
}

File: asg3-java-matrix/List.java
/**
 * @author Tz-Shiuan Lin
 * ID: 1411593
 * List.java
 * CS101-pa3
 */

public class List {

	private class Node {
		private Object data;
		private Node beforeNode;
		private Node afterNode;

		Node(Object data2Insert) {
			this.data = data2Insert;
			this.beforeNode = this.afterNode = null;
		}

		public String toString() {
			return data.toString();
		}

		Object getData() {
			return data;
		}

		Node getBeforeNode() {
			return beforeNode;
		}

		Node getAfterNode() {
			return afterNode;
		}

		void setBeforeNode(Node beforeNode) {
			this.beforeNode = beforeNode;
		}

		void setAfterNode(Node afterNode) {
			this.afterNode = afterNode;
		}
	}

	private Node frontNode;
	private Node backNode;
	private Node cursorNode;
	private int length;
	private int index;

	/**
	 * This is the constructor which creates a new empty list instance.
	 */
	public List() {
		this.frontNode = this.backNode = this.cursorNode = null;
		setLength(0);
		setIndex(-1);
	}

	/* ACCESS FUNCTIONS */
	/**
	 * This method returns the length of the list.
	 * 
	 * @note The key word 'front' represents first element (index '0'), and
	 *       'back' represents last index (length() - 1). See {@link #front()}
	 *       and {@link #back()}
	 * @return number of elements in the list.
	 */
	int length() {
		return this.length;
	}

	/**
	 * If a cursor is defined, this method returns the cursor's index. This
	 * method returns -1 if the cursor is undefined.
	 * 
	 * @return index of the cursor or -1 if cursor is undefined
	 */
	int index() {
		return this.index;
	}

	/**
	 * This method returns the front element (index 0 element).
	 * 
	 * When the list is empty, {@link #length()} should return 0 and this method
	 * should throw an exception.
	 * 
	 * @throws RuntimeException
	 *             when the list is empty ({@link #length()} return 0)
	 * 
	 * @return the front element in the list.
	 */
	Object front() throws RuntimeException {
		if (length() > 0) {
			return getFrontNode().getData();
		} else {
			throw new RuntimeException("List is empty");
		}
	}

	/**
	 * This method returns the front element (index {@link #index()} - 1
	 * element).
	 * 
	 * When the list is empty, {@link #length()} should return 0 and this method
	 * should throw RuntimeException.
	 * 
	 * @throws RuntimeException
	 *             when the list is empty ({@link #length()} return 0)
	 * 
	 * @return the last element in the list.
	 */
	Object back() throws RuntimeException {
		if (length() > 0) {
			return getBackNode().getData();
		} else {
			throw new RuntimeException("List is empty");
		}
	}

	/**
	 * This method returns the cursor's current element.
	 * 
	 * @throws RuntimeException
	 *             when the list is empty ({@link #length()} returns 0) or
	 *             cursor is undefined ( {@link #index()} returns -1)
	 * @return cursor's current element
	 * 
	 */
	Object get() throws RuntimeException {
		if (length() > 0 && index() >= 0) {
			return getCursorNode().getData();
		} else {
			throw new RuntimeException("Cursor is not defined");
		}
	}

	/**
	 * This method compares 2 lists and return true if the integer element in
	 * the sequence are the same sequence.
	 * 
	 * @param Object
	 *            toCompare Another list to compare with. Object that is not a
	 *            list will return false.
	 * @return true if 2 list contain same sequence. false otherwise.
	 */
	public boolean equals(Object toCompare) {
		if (toCompare instanceof List) {
			List list2Compare = (List) toCompare;
			if (this.length() != list2Compare.length()) {
				return false;
			}

			this.moveFront();
			list2Compare.moveFront();
			while (this.index() != -1) {
				if (this.cursorNode.getData().equals(
						list2Compare.cursorNode.getData())) {
					this.moveNext();
					list2Compare.moveNext();
				} else {
					return false;
				}
			}
			return true;
		} else {
			return false;
		}
	}
	
	/* MANIPULATION PROCEDURES */	
	/**
	 * This method resets the list to its empty state.
	 */
	void clear() {
		this.cursorNode = this.frontNode = this.backNode = null;
		this.setLength(0);
		this.setIndex(-1);
	}

	/**
	 * This method places the cursor under the front element. If the list is
	 * empty, this method does nothing.
	 */
	void moveFront() {
		if (this.length() > 0) {
			this.setCursorNode(this.getFrontNode());
			this.setIndex(0);
		}
	}

	/**
	 * This method places the cursor under the back element. If the list is
	 * empty, this method does nothing.
	 */
	void moveBack() {
		if (this.length() > 0) {
			this.setCursorNode(this.getBackNode());
			this.setIndex(this.length() - 1);
		}
	}

	/**
	 * This method moves the cursor one step closer to the front element. This
	 * method sets the cursor to undefined state if the cursor is defined at the
	 * front element. If the list is empty or the cursor is undefined, then this
	 * method does nothing.
	 */
	void movePrev() {
		if (this.length() > 0 && this.index() > -1) {
			if (this.index() == 0) {
				this.cursorNode = null;
				this.setIndex(-1);
			} else {
				this.cursorNode = this.cursorNode.getBeforeNode();
				this.decrementIndex();
			}
		}
	}

	/**
	 * This method moves the cursor one step away to the front element. This
	 * method sets the cursor to undefined state if the cursor is defined at the
	 * back element. If the list is empty or the cursor is undefined, then this
	 * method does nothing.
	 */
	void moveNext() {
		if (this.length() > 0 && this.index() > -1) {
			if (this.index() == this.length() - 1) {
				this.cursorNode = null;
				this.setIndex(-1);
			} else {
				this.cursorNode = this.cursorNode.getAfterNode();
				this.incrementIndex();
			}
		}
	}

	/**
	 * This method inserts a new data into the list, and makes the data the
	 * front element.
	 * 
	 * @note if the cursor is selecting front element, then this method sets the
	 *       cursorNode to the new front element or increments the index
	 *       otherwise.
	 * 
	 * @param data
	 *            Object to be inserted
	 */
	void prepend(Object data) {
		this.incrementLength();
		Node newFront = new Node(data);
		newFront.setBeforeNode(null);
		newFront.setAfterNode(this.getFrontNode());
		try {
			this.getFrontNode().setBeforeNode(newFront);
		} catch (Exception e) {
			// The list is probably empty
			this.setBackNode(newFront);
			// TODO check if any action are required
		}
		this.setFrontNode(newFront);
		if (this.index() == 0) {
			this.setCursorNode(newFront);
		} else if (this.index() > 0) {
			this.incrementIndex();
		}

	}

	/**
	 * This method inserts a new data into the list, and makes the data the back
	 * element. (Insert at the back)
	 * 
	 * @param data
	 *            Object to be inserted
	 */
	void append(Object data) {
		this.incrementLength();
		Node newBack = new Node(data);
		newBack.setAfterNode(null); //
		newBack.setBeforeNode(this.getBackNode());
		try {
			this.getBackNode().setAfterNode(newBack);
		} catch (NullPointerException e) {
			// List is probably empty
			this.setFrontNode(newBack);
			// TODO check if any action is required
		}
		this.setBackNode(newBack);
	}

	/**
	 * This method inserts a new data into the list before the cursor's element.
	 * 
	 * @throws RuntimeException
	 *             if {@link #index()} returns -1
	 * @param data
	 *            Object to be inserted
	 */
	void insertBefore(Object data) throws RuntimeException {
		if (this.index() == -1) {
			throw new RuntimeException(
					"Attemp to insert before an undefined cursor");
		} else {
			this.incrementLength();
			Node newBefore = new Node(data);
			newBefore.setAfterNode(this.getCursorNode());
			newBefore.setBeforeNode(this.getCursorNode().getBeforeNode());
			try {
				this.getCursorNode().getBeforeNode().setAfterNode(newBefore);
			} catch (Exception e) {
				// Cursor is probably selecting the front element.
				this.setFrontNode(newBefore);
				// TODO check if any action is required
			}
			this.getCursorNode().setBeforeNode(newBefore);
			this.incrementIndex();
		}

	}

	/**
	 * This method inserts a new data into the list after the cursor's element.
	 * 
	 * @throws RuntimeException
	 *             if {@link #index()} returns -1
	 * @param data
	 *            Object to be inserted
	 */
	void insertAfter(Object data) throws RuntimeException {
		if (this.index() == -1) {
			throw new RuntimeException(
					"Attemp to insert after an undefined cursor");
		} else {
			this.incrementLength();
			Node newAfter = new Node(data);
			newAfter.setBeforeNode(this.getCursorNode());
			newAfter.setAfterNode(this.getCursorNode().getAfterNode());
			try {
				this.getCursorNode().getAfterNode().setBeforeNode(newAfter);
			} catch (Exception e) {
				// Cursor is probably selecting the back element.
				this.setFrontNode(newAfter);
				// TODO check if any action is required
			}
			this.getCursorNode().setAfterNode(newAfter);
		}
	}

	/**
	 * This method removes the front element from the list.
	 * 
	 * @throws RuntimeException
	 *             if {@link #length()} returns 0
	 * @note if the cursor is selecting the front element, then the cursor will
	 *       be set to undefined.
	 */
	void deleteFront() {
		this.decrementLength();
		this.getFrontNode().getAfterNode().setBeforeNode(null);
		this.setFrontNode(this.getFrontNode().getAfterNode());
		if (this.index() == 0) {
			this.setIndex(-1);
			this.setCursorNode(null);
		}
	}

	/**
	 * This method removes the back element from the list.
	 * 
	 * @throws RuntimeException
	 *             if {@link #length()} returns 0
	 * @note if the cursor is selecting the back element, then the cursor will
	 *       be set to undefined.
	 */

	void deleteBack() {
		if (this.index() == this.length() - 1) {
			this.setIndex(-1);
			this.setCursorNode(null);
		}
		this.decrementLength();
		this.getBackNode().getBeforeNode().setAfterNode(null);
		this.setBackNode(this.getBackNode().getBeforeNode());
	}

	/**
	 * This method removes the cursor's element from the list, then the cursor
	 * will be set to undefined.
	 * 
	 * @throws RuntimeException
	 *             if {@link #index()} returns -1
	 * @note if the cursor is selecting the back element, then cursor will be
	 *       set to undefined.
	 */
	void delete() throws RuntimeException {
		if (this.index() > -1) {
			this.decrementLength();
			this.setIndex(-1);
			Node before = this.getCursorNode().getBeforeNode();
			Node after = this.getCursorNode().getAfterNode();
			try {
				before.setAfterNode(after);
			} catch (Exception e) {
				setFrontNode(after);
			}
			try {
				after.setBeforeNode(before);
			} catch (Exception e) {
				setBackNode(before);			}
			this.setCursorNode(null);
		} else {
			throw new RuntimeException("Attemp to delete an undefined cursor");
		}

	}

	/* OTHER METHODS */
	public String toString() {
		Node temp = getFrontNode();
		String message = "";
		while (temp != null) {
			message += temp.getData().toString();
			message += " ";
			temp = temp.getAfterNode();
		}
		return message;
	}

	/* GETTERS AND SETTERS */
	public Node getFrontNode() {
		return frontNode;
	}

	public Node getBackNode() {
		return backNode;
	}

	public Node getCursorNode() {
		return cursorNode;
	}

	private void setFrontNode(Node frontNode) {
		this.frontNode = frontNode;
	}

	private void setBackNode(Node backNode) {
		this.backNode = backNode;
	}

	private void setCursorNode(Node cursorNode) {
		this.cursorNode = cursorNode;
	}

	private void incrementLength() {
		this.length++;
	}

	private void decrementLength() {
		this.length--;
	}

	public void setLength(int lenght) {
		this.length = lenght;
	}

	private void incrementIndex() {
		this.index++;
	}

	private void decrementIndex() {
		this.index--;
	}

	private void setIndex(int index) {
		this.index = index;
	}
}